sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -52,280 +59,491 @@
+          }
+
+          @Override
+          public State waitUntilFinish(final Duration duration) {
+            return null;
+          }
+
+          @Override
+          public State waitUntilFinish() {
+            return null;
+          }
+
+          @Override
+          public <T> AggregatorValues<T> getAggregatorValues(final Aggregator<?, T> aggregator)
+              throws AggregatorRetrievalException {
+            return null;
+          }
+
+          @Override
+          public MetricResults metrics() {
+            return null;
+          }
+        };
+      }
     }
-  }
 
-  private static class DummyRunner extends PipelineRunner<PipelineResult> {
+    @SuppressWarnings("UnusedReturnValue")
+    private static PCollection<String> addTransform(final PCollection<String> pCollection) {
+      return pCollection.apply("Map2",
+                               MapElements.via(new SimpleFunction<String, String>() {
+
+                                 @Override
+                                 public String apply(final String input) {
+                                   return WHATEVER;
+                                 }
+                               }));
+    }
 
-    @SuppressWarnings("unused") // used by reflection
-    public static DummyRunner fromOptions(final PipelineOptions opts) {
-      return new DummyRunner();
+    private static PCollection<String> pCollection(final Pipeline pipeline) {
+      return pipeline.apply("Create",
+                            Create.of(WORDS).withCoder(StringUtf8Coder.of()))
+                     .apply("Map1",
+                            MapElements.via(new SimpleFunction<String, String>() {
+
+                              @Override
+                              public String apply(final String input) {
+                                return WHATEVER;
+                              }
+                            }));
     }
 
-    @Override
-    public PipelineResult run(final Pipeline pipeline) {
-      return new PipelineResult() {
+    /**
+     * Tests for {@link TestPipeline}s with a non {@link CrashingRunner}.
+     */
+    public static class NonCrashingRunner {
 
-        @Override
-        public State getState() {
-          return null;
-        }
+      private static PipelineOptions dummyRunner() {
+        final PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+        pipelineOptions.setRunner(DummyRunner.class);
+        return pipelineOptions;
+      }
 
-        @Override
-        public State cancel() throws IOException {
-          return null;
-        }
+      @Rule
+      public final transient TestPipeline pipeline = TestPipeline.fromOptions(dummyRunner());
 
-        @Override
-        public State waitUntilFinish(final Duration duration) {
-          return null;
-        }
+      @Category(RunnableOnService.class)
+      @Test
+      public void testNormalFlow_runnableOnService_expectNoException() throws Exception {
```

Comment:
Sorry I did not provide more context here.
`NonCrachingRunner` and `WithCrachingRunner` are kind of meta-tests here. Since I need to test the `TestPipeline` rule in various scenarios, including with and without `@NeedsRunner/RunnableOnService` I use `JUnitCore` to run these tests myself (since otherwise they use the `pom` definitions). In a way, I create a test bed for tests since I would like to test things by writing "user scenarios".

I use this funny looking naming to make the test-bed generic and be able to specify what's the expected behaviour, see also `runTest`.